tests/interlocked_win32_ut/interlocked_win32_ut.c

```diff
@@ -0,0 +1,708 @@
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+
+    umock_c_reset_all_calls();
+}
+
+TEST_FUNCTION_CLEANUP(cleans)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_001: [interlocked_add shall call InterlockedAdd from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_002 : [interlocked_add shall return the result of the addition.]*/
+TEST_FUNCTION(interlocked_add_calls_InterlockedAdd)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MAX;
+    int32_t value = INT32_MIN;
+    STRICT_EXPECTED_CALL(mock_InterlockedAdd((volatile LONG*)&addend, (LONG)value))
+        .SetReturn(-1);
+
+    ///act
+    int32_t return_val = interlocked_add(&addend, value);
+
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int32_t, -1, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_003: [interlocked_and shall call InterlockedAnd from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_004 : [interlocked_and shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_and_calls_InterlockedAnd)
+{
+    ///arrange
+    volatile uint32_t destination = 0xF0F0F0F0;
+    uint32_t value = 0x0F0F0F0F;
+    STRICT_EXPECTED_CALL(mock_InterlockedAnd((volatile LONG*)&destination, (LONG)value))
+        .SetReturn(0xF0F0F0F0);
+
+    ///act
+    uint32_t return_val = (uint32_t)interlocked_and((volatile int32_t*)&destination, (int32_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(uint32_t, 0xF0F0F0F0, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_005: [interlocked_and_16 shall call InterlockedAnd16 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_006 : [interlocked_and_16 shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_and_16_calls_InterlockedAnd16)
+{
+    ///arrange
+    volatile uint16_t destination = 0xF0F0;
+    uint16_t value = 0x0F0F;
+    STRICT_EXPECTED_CALL(mock_InterlockedAnd16((volatile SHORT*)&destination, (SHORT)value))
+        .SetReturn(0xF0F0);
+
+    ///act
+    uint16_t return_val = (uint16_t)interlocked_and_16((volatile int16_t*)&destination, (int16_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(uint16_t, 0xF0F0, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_007: [interlocked_and_64 shall call InterlockedAnd64 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_008 : [interlocked_and_64 shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_and_64_calls_InterlockedAnd64)
+{
+    ///arrange
+    volatile uint64_t destination = 0xF0F0F0F0F0F0F0F0;
+    uint64_t value = 0x0F0F0F0F0F0F0F0F;
+    STRICT_EXPECTED_CALL(mock_InterlockedAnd64((volatile LONG64*)&destination, (LONG64)value))
```

Comment:
>mock_InterlockedAnd64 [](start = 25, length = 21)

mock_InterlockedAnd64 seems to return some LONG64. Which is signed.

0xf0f0... is that meant to be the representation of some negative value?

Reply:
this is spefically about .SetReturn(...)

---
In reply to: [427743537](https://github.com/Azure/azure-c-util/pull/43#discussion_r427743537) [](ancestors = 427743537)

Reply:
Yes, that is a negative value.

---
In reply to: [427743632](https://github.com/Azure/azure-c-util/pull/43#discussion_r427743632) [](ancestors = 427743632,427743537)